processor/schemaprocessor/internal/translation/modifier.go

```diff
@@ -0,0 +1,194 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package translation // import "github.com/open-telemetry/opentelemetry-collector-contrib/processor/schemaprocessor/internal/translation"
+
+import (
+	"go.opentelemetry.io/collector/pdata/pcommon"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/schemaprocessor/internal/alias"
+)
+
+// Modifier abstracts a verion change that allows
+// for the changes to be updated or revert.
+type Modifier interface {
+	// UpdateAttrs modifies all matching attributes
+	// names with the configured changes.
+	UpdateAttrs(attrs pcommon.Map)
+	// RevertAttrs modifies all matching attributes
+	// names with the configured changes.
+	RevertAttrs(attrs pcommon.Map)
+	// UpdateAttrsIf only applies the changes configured
+	// if the changes explicitly require the match value
+	// or if there is no configured values to match against.
+	UpdateAttrsIf(match string, attrs pcommon.Map)
+	// RevertAttrsIf only applies the changes configured
+	// if the changes explicitly require the match value
+	// or if there is no configured values to match against.
+	RevertAttrsIf(match string, attrs pcommon.Map)
+	// UpdateSignal will update the name of the signal
+	// if there is a known transition
+	UpdateSignal(signal alias.Signal)
+	// RevertSignal will update the name of the signal
+	// if there is a known transition
+	RevertSignal(signal alias.Signal)
+}
+
+// modify is one change that can be made to a signal
+type modify struct {
+	names map[alias.SignalName]alias.SignalName
+
+	// AppliesTo is used to scope which signals are to be updated.
+	// In the event that this is empty, then all signals are in scope.
+	appliesTo map[alias.SignalName]struct{}
+	attrs     map[alias.AttributeKey]alias.AttributeKey
+}
+
+// modifications is used when there is sets of changes
+// that can not be merged together due to dependant changes
+// on signal name
+type modifications []*modify
+
+var (
+	_ Modifier = (*modify)(nil)
+	_ Modifier = (*modifications)(nil)
+)
+
+func newModify() *modify {
+	return &modify{
+		names:     make(map[string]string),
+		appliesTo: make(map[string]struct{}),
+		attrs:     make(map[alias.AttributeKey]alias.AttributeKey),
+	}
+}
+
+func (mod *modify) UpdateAttrs(attrs pcommon.Map) {
+	for from, to := range mod.attrs {
```

Comment:
I am curious if iterating by `attrs` may result in better performance. Loolkup of `attrs.Get()` likely is slower than `mod.attrs[]` lookup, right, especially if there are a lot of attributes? I am curious how performance of this compares:

```go
attrs.Range(func f(k,v) {
  if v, match := mod.attrs[k]; match {
     ...
  }
})
```

We may also need to verify that the attribute we convert to does not already exist (and error if it does?). 

Also I wonder if we need to make the code work correctly if there are renaming loops (e.g. a->b, b->a). The expectation is that the attributes should be swapped, but the current code will not work correctly in that case. Perhaps we should prohibit that in the spec, currently schemas don't say anything about it.

Finally, since iteration order of maps is not defined, the effects of this function are non-deterministic if there are any renaming loops.

Reply:
Potentially, I did forget that `pcommon.Map` is not a conventional golang map so there is a cost there that I would need to come back to.

> We may also need to verify that the attribute we convert to does not already exist (and error if it does?).

That is an interesting question / point, I've not addressed the concern that there is a missing attribute that is defined within the schema. Is there an expected behaviour that is defined within the spec?

> Also I wonder if we need to make the code work correctly if there are renaming loops (e.g. a->b, b->a). The expectation is that the attributes should be swapped, but the current code will not work correctly in that case. Perhaps we should prohibit that in the spec, currently schemas don't say anything about it.

Is this an expected behaviour? I believe I have a test in either this or the next PR that looks to address this in a test where a signal revision has `a --> b` and `b --> a`. Since it only performs one modification per scheme revision. However, I think if you list several modifications within one schema then it could potential happen?
I would need to review this again.

> Finally, since iteration order of maps is not defined, the effects of this function are non-deterministic if there are any renaming loops.

Did you have a inclination of what you'd expect in this scenario? 
My initial thought would be to ignore them and have it so an attribute can only be modified once per revision but I am not fixed to the idea.